pkg/store/s3/s3.go

```diff
@@ -0,0 +1,146 @@
+package s3
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/nrkno/terraform-registry/pkg/core"
+	"go.uber.org/zap"
+)
+
+// S3API defines the subset of S3 client methods used by S3Store
+type S3API interface {
+	ListObjectsV2WithContext(ctx aws.Context, input *s3.ListObjectsV2Input) (*s3.ListObjectsV2Output, error)
+	HeadObjectWithContext(ctx aws.Context, input *s3.HeadObjectInput) (*s3.HeadObjectOutput, error)
+}
+
+// S3StoreInterface defines the interface for S3Store
+type S3StoreInterface interface {
+	ListModuleVersions(ctx context.Context, namespace, name, system string) ([]*core.ModuleVersion, error)
+	GetModuleVersion(ctx context.Context, namespace, name, system, version string) (*core.ModuleVersion, error)
+}
+
+// S3Store implements S3StoreInterface
+type S3Store struct {
+	client s3iface.S3API
+	cache  map[string][]*core.ModuleVersion
+	region string
+	bucket string
+	logger *zap.Logger
+	mut    sync.RWMutex
+}
+
+func NewS3Store(client s3iface.S3API, region string, bucket string, logger *zap.Logger) *S3Store {
+	if logger == nil {
+		logger = zap.NewNop()
+	}
+
+	return &S3Store{
+		client: client,
+		cache:  make(map[string][]*core.ModuleVersion),
+		region: region,
+		bucket: bucket,
+		logger: logger,
+	}
+}
+
+func (s *S3Store) ListModuleVersions(ctx context.Context, namespace, name, system string) ([]*core.ModuleVersion, error) {
+	s.mut.RLock()
+	defer s.mut.RUnlock()
+
+	addr := filepath.Join(namespace, name, system)
+	vers, err := s.fetchModuleVersions(ctx, addr)
+	if err != nil {
+		return nil, err
+	}
+
+	return vers, nil
+}
+
+func (s *S3Store) fetchModuleVersions(ctx context.Context, address string) ([]*core.ModuleVersion, error) {
+	p := address + "/"
+	in := &s3.ListObjectsV2Input{
+		Bucket: aws.String(s.bucket),
+		Prefix: aws.String(p),
+	}
+	out, err := s.client.ListObjectsV2WithContext(ctx, in)
+	if err != nil {
+		return nil, err
+	}
+
+	vers := make([]*core.ModuleVersion, 0)
+	for _, o := range out.Contents {
+		path := o.Key
+		if isValidModuleSourcePath(*path) {
+			vers = append(vers, &core.ModuleVersion{
+				Version:   strings.Split(*path, "/")[3],
+				SourceURL: fmt.Sprintf("s3::https://%s.s3.%s.amazonaws.com/%s", s.bucket, s.region, *path),
+			})
+		}
+	}
+
+	s.cache[address] = vers
+
+	return vers, nil
+}
+
+func (s *S3Store) GetModuleVersion(ctx context.Context, namespace, name, system, version string) (*core.ModuleVersion, error) {
+	s.mut.RLock()
```

Comment:
Synchronisation should be moved to the `fetchModuleVersion` where it's actually needed, instead of being indirectly controlled from here.

Reply:
A ReadLock is being used when the cache is actually being written to in certain code paths on line 135.

Reply:
This appears fixed!